### ISSUE TITLE
Automatically exclude .gitignore content

### ISSUE THREAD

**Issue by aheritier** [openrewrite/rewrite-maven-plugin#909]
## What problem are you trying to solve?

Our project use Git as SCM (classical) and has a `.gitignore` file to exclude some files and folders  which shouldn't be store in the SCM (often large folders like node_modules etc). OpenRewrite proposes a [manual option to exclude some folders](https://docs.openrewrite.org/reference/faq#is-it-possible-to-update-only-a-few-files-with-a-recipe-rather-than-all-of-them) from the update. By default, or at least optionally it should allow to follow the .gitignore content because it is 99.99999% sure that you never want to update them (whatever are the update rules in the recipes).

## Describe the solution you'd like

Either by default and in any case with an option to turn on or off the feature to respect the .gitignore content

## Have you considered any alternatives or workarounds?

For now I manually add exclusion entries to avoid to scan the whole project.

**Comment by aheritier** [comment]
It makes sense to me @timtebeek and thanks for your feedbacks @shanman190
I effectively use the maven plugin (not sure why ;-D ) and based on the exchange here I fill that what is the most "annoying" for me is that the plugin doesn't to configure easily what should be excluded from parsing or from updating steps. There is only the [`exclusions`](https://openrewrite.github.io/rewrite-maven-plugin/run-mojo.html#exclusions) parameter.

**Comment by timtebeek** [comment]
Yes I think we can do better; for comparison, [the Gradle project parser does use the OmniParser](https://github.com/openrewrite/rewrite-gradle-plugin/blob/e1ecb67503e3b0f59fa4f013bb7d76fa3789673a/plugin/src/main/java/org/openrewrite/gradle/isolated/DefaultProjectParser.java#L1179-L1191), and skips over already parsed files, as well as anything that is Git ignored. From the looks of it, most of what's in [the Maven plugin's ResourceParser](https://github.com/openrewrite/rewrite-maven-plugin/blob/9a0d4fce38efd0d16c72b76beab60a5e0f81f3d6/src/main/java/org/openrewrite/maven/ResourceParser.java#L42) can move to use the OmniParser to gain that git ignore most easily. Something to look into, if I find the time.

**Comment by aheritier** [comment]
Thanks Tim. It's not blocker/urgent. I might try to look at it but it won't be before 6 months.

**Comment by timtebeek** [comment]
Bad bot! We'll continue to keep this open, and I hope all is well there! :)

**Comment by timtebeek** [comment]
This has been implemented. The plugin now automatically excludes files that are both gitignored and untracked via `GitIgnore.isIgnoredAndUntracked()` in `MavenMojoProjectParser.isExcluded()`, so directories like `node_modules` are skipped without manual configuration.
